File: Stage1st/Manager/Navigation/NavigationControllerDelegate.m
//
//  NavigationControllerDelegate.m
//  Stage1st
//
//  Created by Zheng Li on 11/16/14.
//  Copyright (c) 2014 Renaissance. All rights reserved.
//

#import "NavigationControllerDelegate.h"

#define _TRIGGER_THRESHOLD 60.0f
#define _TRIGGER_VELOCITY_THRESHOLD 500.0f
#define _COLOR_CHANGE_TRIGGER_THRESHOLD 100.0f

@interface NavigationControllerDelegate () <UIGestureRecognizerDelegate>

@property (strong, nonatomic) S1Animator *popAnimator;
@property (strong, nonatomic) S1Animator *pushAnimator;
@property (strong, nonatomic) UIPercentDrivenInteractiveTransition *interactionController;

@end

@implementation NavigationControllerDelegate

- (instancetype)initWithNavigationController:(S1NavigationViewController *)navigationController
{
    self = [super init];
    if (self != nil) {
        self.navigationController = navigationController;
        self.panRecognizer = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(pan:)];
        self.panRecognizer.delegate = self;
        [self.navigationController.view addGestureRecognizer:self.panRecognizer];

        self.popAnimator = [[S1Animator alloc] initWithDirection:TransitionDirectionPop];
        self.pushAnimator = [[S1Animator alloc] initWithDirection:TransitionDirectionPush];
    }
    return self;
}

- (void)colorPan:(UIPanGestureRecognizer *)recognizer
{
    UIView *view = self.navigationController.view;
    CGPoint translation = [recognizer translationInView:view];
    if (recognizer.state == UIGestureRecognizerStateChanged) {
        DDLogVerbose(@"[ColorPan] %f",translation.y);
        if (translation.y > _COLOR_CHANGE_TRIGGER_THRESHOLD && [[NSUserDefaults standardUserDefaults] boolForKey:@"NightMode"] == NO) {
            [[NSUserDefaults standardUserDefaults] setBool:YES forKey:@"NightMode"];
            [[ColorManager shared] switchPalette:PaletteTypeNight];
            recognizer.enabled = NO;
            recognizer.enabled = YES;
        }
        if (translation.y < -_COLOR_CHANGE_TRIGGER_THRESHOLD && [[NSUserDefaults standardUserDefaults] boolForKey:@"NightMode"] == YES) {
            [[NSUserDefaults standardUserDefaults] setBool:NO forKey:@"NightMode"];
            [[ColorManager shared] switchPalette:PaletteTypeDay];
            recognizer.enabled = NO;
            recognizer.enabled = YES;
        }
    }
}

- (void)pan:(UIPanGestureRecognizer*)recognizer
{
    UIView* view = self.navigationController.view;
    CGPoint translation = [recognizer translationInView:view];
    if (recognizer.state == UIGestureRecognizerStateBegan) {
        self.interactionController = [UIPercentDrivenInteractiveTransition new];
        self.interactionController.completionCurve = UIViewAnimationCurveLinear;
        self.popAnimator.curve = UIViewAnimationOptionCurveLinear;
        [self.navigationController popViewControllerAnimated:YES];
        DDLogVerbose(@"[NavPan] Begin.");
    }
    else if (recognizer.state == UIGestureRecognizerStateChanged) {
        CGFloat windowWidth = UIApplication.sharedApplication.delegate.window.bounds.size.width;
        [self.interactionController updateInteractiveTransition:translation.x > 0 ? translation.x / windowWidth : 0];
        DDLogVerbose(@"[NavPan] Update: %f.", translation.x > 0 ? translation.x / windowWidth : 0);
    } else if (recognizer.state == UIGestureRecognizerStateEnded) {
        CGFloat velocityX = [recognizer velocityInView:view].x;
        CGRect screenRect = [[UIScreen mainScreen] bounds];
        CGFloat screenWidth = screenRect.size.width;
        if ((translation.x > _TRIGGER_THRESHOLD || velocityX > _TRIGGER_VELOCITY_THRESHOLD) && velocityX >= -100) {
            self.interactionController.completionSpeed = 0.3 / fmin((screenWidth - fmin(translation.x, 0)) / fabs(velocityX), 0.3);
            DDLogVerbose(@"[NavPan] Finish Speed: %f", self.interactionController.completionSpeed);
            [self.interactionController finishInteractiveTransition];
            if (self.navigationController.viewControllers.count == 1) {
                self.panRecognizer.enabled = NO;
            }
        } else {
            self.interactionController.completionSpeed = 0.3 / fmin(fabs(translation.x / velocityX), 0.3);
            DDLogVerbose(@"[NavPan] Cancel Speed: %f", self.interactionController.completionSpeed);
            [self.interactionController cancelInteractiveTransition];
        }
        self.interactionController = nil;
        self.popAnimator.curve = UIViewAnimationOptionCurveEaseInOut;
    } else {
        DDLogError(@"[NavPan] Other Interaction Event:%ld", (long)recognizer.state);
        [self.interactionController cancelInteractiveTransition];
        self.interactionController = nil;
        self.popAnimator.curve = UIViewAnimationOptionCurveEaseInOut;
    }
}

#pragma mark - UIGestureRecognizerDelegate

- (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer
{
    if (gestureRecognizer == self.colorPanRecognizer) {
        if (gestureRecognizer.numberOfTouches == 1) {
            return NO;
        } else {
            return YES;
        }
    }

    if (gestureRecognizer == self.panRecognizer) {
        if (self.navigationController.viewControllers.count == 1) {
            return NO;
        }
    }
    return YES;
}

- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer
{
    if (gestureRecognizer == self.colorPanRecognizer) {
//        DDLogVerbose(@"pan gesture simultaneously with %@", otherGestureRecognizer);
//        if ([otherGestureRecognizer.view isKindOfClass:[NSClassFromString(@"WKScrollView") class]] && [otherGestureRecognizer isKindOfClass:[NSClassFromString(@"UIScrollViewPanGestureRecognizer") class]]) {
//            return YES;
//        }
    }

    return NO;
}

#pragma mark - UINavigationControllerDelegate

- (id<UIViewControllerAnimatedTransitioning>)navigationController:(UINavigationController *)navigationController
                                  animationControllerForOperation:(UINavigationControllerOperation)operation
                                               fromViewController:(UIViewController *)fromVC
                                                 toViewController:(UIViewController *)toVC
{
    if (operation == UINavigationControllerOperationPop) {
        return self.popAnimator;
    } else if (operation == UINavigationControllerOperationPush) {
        if (self.navigationController.viewControllers.count > 1) {
            self.panRecognizer.enabled = YES;
        }
        return self.pushAnimator;
    }
    return nil;
}

- (id<UIViewControllerInteractiveTransitioning>)navigationController:(UINavigationController *)navigationController
                         interactionControllerForAnimationController:(id<UIViewControllerAnimatedTransitioning>)animationController
{
    return self.interactionController;
}

- (UIInterfaceOrientationMask)navigationControllerSupportedInterfaceOrientations:(UINavigationController *)navigationController
{
    if ([[NSUserDefaults standardUserDefaults] boolForKey:@"ForcePortraitForPhone"]) {
        if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone) {
            return UIInterfaceOrientationMaskPortrait;
        }
    }
    return UIInterfaceOrientationMaskAll;
}

@end

File: Stage1st/Manager/Navigation/NavigationControllerDelegate.swift
//
//  NavigationControllerDelegate.swift
//  Stage1st
//
//  Created by Zheng Li on 2019/1/20.
//  Copyright © 2019 Renaissance. All rights reserved.
//

import UIKit

class NavigationControllerDelegate: NSObject {

    weak var navigationController: UINavigationController?
    lazy var panRecognizer: UIPanGestureRecognizer = { UIPanGestureRecognizer(target: self, action: #selector(pan(_:))) }()

    private var interactionController: UIPercentDrivenInteractiveTransition?

    init(navigationController: UINavigationController) {
        self.navigationController = navigationController

        super.init()

        panRecognizer.delegate = self
        navigationController.view.addGestureRecognizer(panRecognizer)
    }
}

// MARK: - Action

extension NavigationControllerDelegate {

    @objc func pan(_ gestureRecognizer: UIPanGestureRecognizer) {
        guard let view = gestureRecognizer.view else {
            assert(false, "this should never happen!")
            S1LogError("this should never happen!")
            return
        }

        let translation = gestureRecognizer.translation(in: view)

        switch gestureRecognizer.state {
        case .began:
            interactionController = UIPercentDrivenInteractiveTransition()
            navigationController?.popViewController(animated: true)
            S1LogDebug("Begin")
        case .changed:
            let width = view.bounds.width
            let percent = max(translation.x / width, 0.0)
            interactionController?.update(percent)
            S1LogVerbose("Update: \(percent)")
        case .ended:
            let width = view.bounds.width
            let velocityX = gestureRecognizer.velocity(in: view).x

            let triggerThreshold: CGFloat = 60.0
            let triggerVelocityThreshold: CGFloat = 500.0

            let shouldFinish = (translation.x > triggerThreshold && velocityX >= -100.0) || velocityX > triggerVelocityThreshold

            if shouldFinish {
                let timeLeft = (width - min(translation.x, 0.0)) / abs(velocityX)
                interactionController?.completionSpeed = 0.3 / min(timeLeft, 0.3)
                interactionController?.finish()
                S1LogDebug("Finish with speed: \(String(describing: interactionController?.completionSpeed))")
            } else {
                let timeUsed = abs(translation.x / velocityX)
                interactionController?.completionSpeed = 0.3 / min(timeUsed, 0.3)
                interactionController?.cancel()
                S1LogDebug("Cancel with speed: \(String(describing: interactionController?.completionSpeed))")
            }

            interactionController = nil
        case .possible:
            S1LogDebug("Possible")
        case .failed, .cancelled:
            S1LogDebug("\(gestureRecognizer.state.rawValue)")
            interactionController?.cancel()
            interactionController = nil
        @unknown default:
            break
        }
    }
}

// MARK: - UIGestureRecognizerDelegate

extension NavigationControllerDelegate: UIGestureRecognizerDelegate {

    func gestureRecognizerShouldBegin(_ gestureRecognizer: UIGestureRecognizer) -> Bool {
        if gestureRecognizer === self.panRecognizer {
            if navigationController?.viewControllers.count != 1 {
                return true
            } else {
                return false
            }
        } else {
            return true
        }
    }
}

// MARK: - UINavigationControllerDelegate

extension NavigationControllerDelegate: UINavigationControllerDelegate {

    func navigationController(
        _ navigationController: UINavigationController,
        animationControllerFor operation: UINavigationController.Operation,
        from fromVC: UIViewController,
        to toVC: UIViewController
    ) -> UIViewControllerAnimatedTransitioning? {
        switch operation {
        case .push:
            return S1Animator(direction: .push)
        case .pop:
            return S1Animator(direction: .pop)
        case .none:
            return nil
        @unknown default:
            return nil
        }
    }

    func navigationController(
        _ navigationController: UINavigationController,
        interactionControllerFor animationController: UIViewControllerAnimatedTransitioning
    ) -> UIViewControllerInteractiveTransitioning? {
        return self.interactionController
    }

    func navigationControllerSupportedInterfaceOrientations(_ navigationController: UINavigationController) -> UIInterfaceOrientationMask {
        if AppEnvironment.current.settings.forcePortraitForPhone.value && UIDevice.current.userInterfaceIdiom == .phone {
            return .portrait
        } else {
            return .all
        }
    }
}
